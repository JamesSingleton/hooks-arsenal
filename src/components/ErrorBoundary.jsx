import { PureComponent } from 'react';
import PropTypes from 'prop-types';

class ErrorBoundary extends PureComponent {
  // eslint-disable-next-line unicorn/prevent-abbreviations
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    return { hasError: true, error };
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    const { handleError } = this.props;
    return handleError(error, errorInfo);
  }

  render() {
    const { hasError, error } = this.state;
    const { render, children, renderErrorText } = this.props;

    // Prevent rendering of children that caused the error, render fallbacks instead
    if (hasError) {
      // Render either components from renderError() or nothing
      return renderErrorText ? renderErrorText({ ...this.props, error }) : null;
    }

    // Render either components from render() or children or null
    return render ? render(this.props) : children || null;
  }
}

ErrorBoundary.propTypes = {
  handleError: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  render: PropTypes.func,
  renderErrorText: PropTypes.func,
};

ErrorBoundary.defaultProps = {
  children: undefined,
  render: undefined,
  renderErrorText: undefined,
};

export default ErrorBoundary;
