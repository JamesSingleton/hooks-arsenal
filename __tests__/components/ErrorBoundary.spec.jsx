import React from 'react';
import { create } from 'react-test-renderer';
import ErrorBoundary from '../../src/components/ErrorBoundary';

describe('ErrorBoundary', () => {
  const handleError = jest.fn();
  const ErrorThrower = ({ errrorMessage = 'uh oh' }) => {
    throw new Error(errrorMessage);
  };

  beforeEach(() => {
    handleError.mockReset();
  });

  it('renders its children', () => {
    const render = create(
      <ErrorBoundary handleError={handleError}>
        <div>I am a child that should render.</div>
      </ErrorBoundary>
    );
    expect(render.toJSON()).toMatchSnapshot();
  });

  it('should render null if an error occurs and no fallback render is provided', () => {
    const render = create(
      <ErrorBoundary handleError={handleError}>
        <ErrorThrower />
      </ErrorBoundary>
    );
    expect(render.toJSON()).toEqual(null);
  });

  it('should render the fallback render if provided and an error occurs', () => {
    const FallbackRender = () => <div>I am a fallback render</div>;
    const render = create(
      <ErrorBoundary handleError={handleError} renderErrorText={FallbackRender}>
        <ErrorThrower />
      </ErrorBoundary>
    );
    expect(render.toJSON()).toMatchSnapshot();
  });

  it('calls the provided error handler should an error occur', () => {
    create(
      <ErrorBoundary handleError={handleError}>
        <ErrorThrower />
      </ErrorBoundary>
    );
    expect(handleError).toHaveBeenCalledTimes(1);
  });

  it('calls the render function if provided', () => {
    const ProvidedRender = () => <div>I will render instead of the children</div>;
    const render = create(
      <ErrorBoundary handleError={handleError} render={ProvidedRender}>
        <div>I will not render because a render function was provided</div>
      </ErrorBoundary>
    );
    expect(render.toJSON()).toMatchSnapshot();
  });
});
