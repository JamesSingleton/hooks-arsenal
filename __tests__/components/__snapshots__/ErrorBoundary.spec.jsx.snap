// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary calls the render function if provided 1`] = `
<div>
  I will render instead of the children
</div>
`;

exports[`ErrorBoundary renders its children 1`] = `
<div>
  I am a child that should render.
</div>
`;

exports[`ErrorBoundary should render the fallback render if provided and an error occurs 1`] = `
<div>
  I am a fallback render
</div>
`;
